File: stack_queue/q_in_q/q_in_q.py
class Queue:
    def __init__(self):
        """
        Initialize an empty queue.
        """
        self.data = []  # List to store the queue elements
        self.front_idx = 0  # Index of the front element

    def enqueue(self, value):
        """
        Add an element to the queue.

        Args:
            value: The value to be added to the queue.
        """
        self.data.append(value)

    def dequeue(self):
        """
        Remove and return the front element of the queue.

        Returns:
            The front element of the queue.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty():
            raise IndexError("Queue underflow")
        value = self.data[self.front_idx]
        self.front_idx += 1
        return value

    def front(self):
        """
        Return the front element without removing it.

        Returns:
            The front element of the queue.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.data[self.front_idx]

    def is_empty(self):
        """
        Check if the queue is empty.

        Returns:
            True if the queue is empty, otherwise False.
        """
        return self.front_idx >= len(self.data)

    def size(self):
        """
        Return the size of the queue.

        Returns:
            The number of elements in the queue.
        """
        return len(self.data) - self.front_idx

    def print(self):
        """
        Print the contents of the queue.
        """
        print(" ".join(map(str, self.data[self.front_idx:])))

    def get_data(self):
        """
        Get the underlying data of the queue.

        Returns:
            The list representing the queue's contents.
        """
        return self.data[self.front_idx:]

    def reset(self):
        """
        Reset the queue's data to remove unused space from the front.
        """
        self.data = self.data[self.front_idx:]
        self.front_idx = 0


def merge_inner_queues_in_queue(outer_queue, i, j, k):
    """
    Merge two inner queues in the outer queue and insert at position `k`.

    Args:
        outer_queue (Queue): The outer queue containing inner queues.
        i (int): Index of the first inner queue to merge.
        j (int): Index of the second inner queue to merge.
        k (int): The position to insert the merged queue.

    Raises:
        IndexError: If indices `i`, `j`, or `k` are invalid.
    """
    if i >= outer_queue.size() or j >= outer_queue.size() or k > outer_queue.size() or i == j:
        raise IndexError("Invalid indices for merging")

    # Convert the queue to a list for easier manipulation
    outer_data = outer_queue.get_data()

    # Create a new queue for the merged data
    merged_queue = Queue()

    # Merge the i-th queue
    for val in outer_data[i].get_data():
        merged_queue.enqueue(val)

    # Merge the j-th queue
    for val in outer_data[j].get_data():
        merged_queue.enqueue(val)

    # Remove i-th and j-th queues
    if i < j:
        del outer_data[j]
        del outer_data[i]
    else:
        del outer_data[i]
        del outer_data[j]

    # Insert the merged queue at position `k`
    outer_data.insert(k, merged_queue)

    # Replace the outer queue with the updated list
    new_outer_queue = Queue()
    for q in outer_data:
        new_outer_queue.enqueue(q)

    outer_queue.data = new_outer_queue.data
    outer_queue.front_idx = new_outer_queue.front_idx


# Main function for testing
if __name__ == "__main__":
    print("=== Testing Queue in Queue ===")

    # Create the outer queue
    outer_queue = Queue()

    # Create and enqueue inner queues
    queue1 = Queue()
    queue1.enqueue(1)
    queue1.enqueue(2)

    queue2 = Queue()
    queue2.enqueue(3)
    queue2.enqueue(4)

    queue3 = Queue()
    queue3.enqueue(5)
    queue3.enqueue(6)

    outer_queue.enqueue(queue1)
    outer_queue.enqueue(queue2)
    outer_queue.enqueue(queue3)

    # Print the outer queue before merging
    print("Outer queue before merging:")
    for q in outer_queue.get_data():
        q.print()

    # Merge queue1 and queue2 and insert at position 1
    merge_inner_queues_in_queue(outer_queue, 0, 1, 1)

    # Print the outer queue after merging
    print("\nOuter queue after merging:")
    for q in outer_queue.get_data():
        q.print()

File: stack_queue/q_in_q/q_in_q.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
using namespace std;

/**********************************************
 * CATALOG OF FUNCTIONS AND CLASSES
 **********************************************

 * GENERIC QUEUE (Queue<T>)
 1. `Queue<T>()`
    - Constructor that initializes an empty queue.
 
 2. `void enqueue(const T& value)`
    - Adds an element `value` to the queue.

 3. `T dequeue()`
    - Removes and returns the front element.
    - Throws a `runtime_error` if the queue is empty.

 4. `T front() const`
    - Returns the front element without removing it.
    - Throws a `runtime_error` if the queue is empty.

 5. `bool isEmpty() const`
    - Checks if the queue is empty.
    - Returns `true` if empty, otherwise `false`.

 6. `size_t size() const`
    - Returns the size of the queue.

 7. `void print() const`
    - Prints the contents of the queue.

 * MERGING QUEUES
 1. `void mergeInnerQueuesInQueue(Queue<Queue<T>>& outerQueue, size_t i, size_t j, size_t k)`
    - Merges two inner queues at indices `i` and `j` in the outer queue.
    - Deletes the original queues at indices `i` and `j`.
    - Inserts the merged queue at position `k`.
 **********************************************/

// Template class for Queue
template <typename T>
class Queue {
private:
    vector<T> data;    // Vector to store the queue elements
    size_t frontIdx;   // Index of the front element
    size_t sizeVal;    // Current size of the queue

public:
    // Constructor
    Queue() : frontIdx(0), sizeVal(0) {}

    // Add an element to the queue
    void enqueue(const T& value) {
        data.push_back(value);
        sizeVal++;
    }

    // Remove and return the front element
    T dequeue() {
        if (isEmpty())
            throw runtime_error("Queue underflow");
        T value = data[frontIdx++];
        sizeVal--;
        return value;
    }

    // Return the front element without removing it
    T front() const {
        if (isEmpty())
            throw runtime_error("Queue is empty");
        return data[frontIdx];
    }

    // Check if the queue is empty
    bool isEmpty() const {
        return sizeVal == 0;
    }

    // Return the size of the queue
    size_t size() const {
        return sizeVal;
    }

    // Print the queue contents
    void print() const {
        for (size_t i = frontIdx; i < data.size(); i++) {
            cout << data[i] << " ";
        }
        cout << endl;
    }

    // Access the underlying data
    vector<T>& getData() {
        return data;
    }
};

// Function to merge two inner queues in a queue and insert at position `k`
template <typename T>
void mergeInnerQueuesInQueue(Queue<Queue<T>>& outerQueue, size_t i, size_t j, size_t k) {
    if (i >= outerQueue.size() || j >= outerQueue.size() || k > outerQueue.size() || i == j) {
        throw runtime_error("Invalid indices for merging");
    }

    // Convert the queue to a vector for easier manipulation
    vector<Queue<T>> outerData = outerQueue.getData();

    // Create a new queue for the merged data
    Queue<T> mergedQueue;

    // Merge the i-th queue
    for (const auto& val : outerData[i].getData()) {
        mergedQueue.enqueue(val);
    }

    // Merge the j-th queue
    for (const auto& val : outerData[j].getData()) {
        mergedQueue.enqueue(val);
    }

    // Erase i-th and j-th queues
    if (i < j) {
        outerData.erase(outerData.begin() + j);
        outerData.erase(outerData.begin() + i);
    } else {
        outerData.erase(outerData.begin() + i);
        outerData.erase(outerData.begin() + j);
    }

    // Insert the merged queue at position `k`
    outerData.insert(outerData.begin() + k, mergedQueue);

    // Replace the outer queue with the updated vector
    Queue<Queue<T>> newOuterQueue;
    for (const auto& q : outerData) {
        newOuterQueue.enqueue(q);
    }
    outerQueue = newOuterQueue;
}

// Main function for testing
int main() {
    cout << "=== Testing Queue in Queue ===" << endl;

    // Create the outer queue
    Queue<Queue<int>> outerQueue;

    // Create and enqueue inner queues
    Queue<int> queue1;
    queue1.enqueue(1);
    queue1.enqueue(2);

    Queue<int> queue2;
    queue2.enqueue(3);
    queue2.enqueue(4);

    Queue<int> queue3;
    queue3.enqueue(5);
    queue3.enqueue(6);

    outerQueue.enqueue(queue1);
    outerQueue.enqueue(queue2);
    outerQueue.enqueue(queue3);

    // Print the outer queue before merging
    cout << "Outer queue before merging:" << endl;
    for (size_t i = 0; i < outerQueue.size(); i++) {
        outerQueue.getData()[i].print();
    }

    // Merge queue1 and queue2 and insert at position 1
    mergeInnerQueuesInQueue(outerQueue, 0, 1, 1);

    // Print the outer queue after merging
    cout << "Outer queue after merging:" << endl;
    for (size_t i = 0; i < outerQueue.size(); i++) {
        outerQueue.getData()[i].print();
    }

    return 0;
}